### PR TITLE
implemented url-shortener ajax integration

### PR DESCRIPTION
**Pull request by celevra**
This PR fixes https://github.com/PrivateBin/PrivateBin/issues/238

## Changes
* added new configuration item urlshortenerajax and description
* added new data attribute to the shortener button and new logic to get the output via ajax

### CODE REVIEW

**Comment by elrido** [comment]
Nice! Simple and minimal change.

File: js/privatebin.js [hunk 1]
```diff
@@ -1195,8 +1195,17 @@ jQuery.PrivateBin = (function($, sjcl, Base64, RawDeflate) {
          */
         function sendToShortener()
         {
-            window.location.href = $shortenButton.data('shortener') +
-                                   encodeURIComponent($pasteUrl.attr('href'));
+            var url = $shortenButton.data('shortener')+encodeURIComponent($pasteUrl.attr('href'));
+
+            if($shortenButton.data('ajax'))
```
Comment (rugk) [inline js/privatebin.js]: Please add the brackets for the if block to improve readability, also for the else block.

File: cfg/conf.sample.php [hunk 2]
```diff
@@ -56,6 +56,12 @@
 ; the pastes encryption key
 ; urlshortener = "https://shortener.example.com/api?link="
 
+; you can set urlshortenerajax to yes if the shortener returns the url plain text
```
Comment (rugk) [inline cfg/conf.sample.php]: maybe rather "true" than "yes"

File: lib/Configuration.php [hunk 3]
```diff
@@ -52,6 +52,7 @@ class Configuration
             'languageselection'        => false,
             'languagedefault'          => '',
             'urlshortener'             => '',
+            'urlshortenerajax'         => '',
```
Comment (rugk) [inline lib/Configuration.php]: default may be "false" or so and not an empty string?

File: cfg/conf.sample.php (hunk 2 above)
Comment (rugk) [reply cfg/conf.sample.php]: also explain that you should point the `urlshortener` to some API endpoint that does this, maybe also link to YOURLS doc for that as we mention it above as an example

File: js/privatebin.js [hunk 4]
```diff
@@ -1195,8 +1195,17 @@ jQuery.PrivateBin = (function($, sjcl, Base64, RawDeflate) {
...
+                $.get(url, function(d) {
+                    $pasteUrl.attr(d);
+                    $pasteUrl.html(d);
```
Comment (rugk) [inline js/privatebin.js]: No, rather not allow the remote server to inject any HTML here. (It just returns text, does not it?)
Please change it to text-only.

**Comment by rugk** [comment]
BTW I've allowed myself to edit the PR text to [let the issue close](https://help.github.com/articles/closing-issues-via-commit-messages/) automatically when this PR is merged.

File: lib/Configuration.php (hunk 3 above)
Comment (celevra) [reply lib/Configuration.php]: data attributes are interpreted as strings so empty is false, everything else is true

File: cfg/conf.sample.php [hunk 5]
```diff
@@ -56,6 +56,12 @@
...
+; YOU HAVE TO set Control-Allow-Origin header on the other side!
```
[the remark this answers is not in the document]
Comment (celevra) [reply cfg/conf.sample.php]: i've changed it, but if someone isn't getting what i mean than i think he/she should not play with Control-Allow-Origin

File: cfg/conf.sample.php (hunk 2 above)
Comment (celevra) [reply cfg/conf.sample.php]: personaly i think "plain text" should be enough, not every body is using YOURLS, i for my self using a self written one.

**Comment by celevra** [comment]
i always read the changes that the code reviewer has done to my pull requests. 
These where only minimal issues and i think it would be faster and for everybody less work if you have simply made the changes that you have request. 
Thinks like brackets and spaces between operators that are only personal style favors and not failures, or a wrong yes instead of a true in the Description don't have to be discussed...
also adding a .text() would have been faster as writing a whole text.

File: js/privatebin.js [hunk 6]
```diff
@@ -1195,17 +1195,21 @@ jQuery.PrivateBin = (function($, sjcl, Base64, RawDeflate) {
...
         {
-            var url = $shortenButton.data('shortener')+encodeURIComponent($pasteUrl.attr('href'));
+            var url = $shortenButton.data('shortener') + encodeURIComponent($pasteUrl.attr('href'));
 
             if($shortenButton.data('ajax'))
+            {
                 $.get(url, function(d) {
                     $pasteUrl.attr(d);
-                    $pasteUrl.html(d);
+                    $pasteUrl.html(d.text());
```
Comment (rugk) [inline js/privatebin.js]: Youz can use `text` directly on the element, instead of converting it three times here… :smile:

**Comment by rugk** [comment]
> Thinks like brackets and spaces between operators that are only personal style favors and not failures

Indeed, they are and it is not bad if you do them in a different way. Of course they are not failures and completely okay to do. Just when contributing to another project, you should adhere to the style they have there. Just because it is also work for the maintainers to adjust such things afterwards.
The ideal case is when you (or any contributor) recognize the code style and try to adapt it. To enable this "learning effect" I also comment for such small things, so when you contribute again, it will likely get through with fewer comments.
And, BTW, it is easier than it sounds, because we have several style checkers enabled [in the project, e.g. EsLint](https://github.com/PrivateBin/PrivateBin/blob/master/.eslintrc), which – when used – instantly give you the feedback I just gave (I hope, at least.).

**Comment by celevra** [comment]
As mentioned before I always read what changes the reviewer does so the "learning effect" is still there. I think js lint can also automatically change such little things.
But I get what you mean and I know that not every contributer reads and tries to understand what the reviewer changes.


________________________________
Von: rugk <notifications@github.com>
Gesendet: Mittwoch, 9. Mai 2018 23:12
An: PrivateBin/PrivateBin
Cc: Philipp Zeitschel; Author
Betreff: Re: [PrivateBin/PrivateBin] implemented url-shortener ajax integration (#313)


Thinks like brackets and spaces between operators that are only personal style favors and not failures

Indeed, they are and it is not bad if you do them in a different way. Of course they are not failures and completely okay to do. Just when contributing to another project, you should adhere to the style they have there. Just because it is also work for the maintainers to adjust such things afterwards.
The ideal case is when you (or any contributor) recognize the code style and try to adapt it. To enable this "learning effect" I also comment for such small things, so when you contribute again, it will likely get through with fewer comments.
And, BTW, it is easier than it sounds, because we have several style checkers enabled in the project, e.g. EsLint<https://github.com/PrivateBin/PrivateBin/blob/master/.eslintrc>, which – when used – instantly give you the feedback I just gave (I hope, at least.).

**Comment by rugk** [comment]
So you actually want HTML support there? Can you tell me an URL shorter, which would actually profit from that? AFAIK (and as far as you describe) they usually just return a plain-text URL.
With which URL shorter did you actually test that, BTW?

Also from a security standpoint ist is not a good idea to just inject some HTML from some other website – it could be self-hosted anyway, yes, but I still feeling uncomfortable with that. Especially if I see no added value in using HTML.

**Comment by celevra** [comment]
hm, eventually i get something wrong?
```
elem.html(d.text())
```
the html function only tells jquery to insert d in the "html" part of the, in this case, a element. 
the .text() function is escaping chars that could be html
and this is what you wanted, we don't trust d so we parse it as text, isn't it?
if i set the element to a just "text" element (i don't exactly understand what you mean here) i can't use the element with html anymore if i want to.

**Comment by rugk** [comment]
Is not `d` the response from the server? The text response?

So while you are at it, you can rename the variable to something more understandable than a single letter. May it be `data`? Or `response` or so…

But I'll have a look at this tomorrow anyway, so I'll have a look at whether what I want works.

**Comment by celevra** [comment]
yes d is data and its absolutely clear as it a parameter of the success function there is absolutely no reason the rename it to anything longer than a simple d

in the meantime i think i know what you mean, you want a 
```
elem.text(d)
```
that is exactly the same as 
```
elem.html(d.text())
```
i will change that

**Comment by rugk** [comment]
It is actually not the same, because the HTML thing may allow HTML code injection. At least when `.html` in jQuery interprets strings as HTML too. (And I think it does.)
So as said, this would be a potential security risk – and, of course, it is an unnecessary conversion of types. (Getting text from server -> "convert" to text -> parse as HTML -> insert into page) The `.text` thing of jQuery is not there.

**Comment by rugk** [comment]
So I tested it a bit and here are the results:
* I used is.gd and lstu.fr for testing.
* First problem: The AJAX request never passed the actual stuff after `#` to the URL shorter, so this is then useless.
* Secondly: The doc (config file here!) needs to mention that you may need to adjust the `connect-src` directive in the CSP header (config: cspheader).
* Thirdly: Many URL shorters do not accept 127.0.0.1 or they do not have CORS headers set, so I actually could not test it completly, so that it works…
* BTW, I also saw that some URL shorters only return JSON. Don't know if we would want special support for that.

**Comment by TommyAsAService** [comment]
Hi guys, can we pick up this PR again? Let me know if I can help.

**Comment by elrido** [comment]
@rugk If I understand your last comment correctly, you are waiting on @celevra to fix some things to get this merged? Could you reiterate what needs to be fixed? At this point the branch is also out of date, so we would need an updated PR for that anyway.

**Comment by elrido** [comment]
This may now have been addressed in #478

**Comment by rugk** [comment]
> This may now have been addressed in #478

But #478 does AFAIK not solve https://github.com/PrivateBin/PrivateBin/issues/238, because that's about an AJAX integration, so you are not redírected to a different site.

**Comment by elrido** [comment]
Sorry, I linked to the wrong merge request: #480 is the one that implements an AJAX call at https://github.com/PrivateBin/PrivateBin/pull/480/files#diff-cf28a0bc390eee2482509aca8212f831R1717

**Comment by rugk** [comment]
Okay, so. So this is replaced by https://github.com/PrivateBin/PrivateBin/pull/480. Closing...

**Pull request closed without merging**